src/main/java/week2/BTree.java

```diff
@@ -0,0 +1,113 @@
+package week2;
+
+
+class BTree<T extends Comparable<T>> {
+
+  private boolean bothFound = false;
+  private Node<T> root;
+
+  BTree(Node<T> root) {
+    this.root = root;
+  }
+
+
+  /* Question 1: Given a Binary Tree and a key, write a function that prints all the ancestors
+                 of the key in the given binary tree.
+                 Example: For the key 5 and the following tree we should print: 3, 9, 16.
+
+     Overview: We define a private method to recursively parse through the tree searching for the
+              element. Once the value is found, the method returns true to signal up to call stack
+               that element is found.
+
+     Time: O(N) where N is the number of elements in tree
+
+   */
+  String printAncestors(T key) {
+    StringBuilder sb = new StringBuilder();
+    printAncestors(key, root, sb);
+
+    return sb.toString();
+  }
+
+  private boolean printAncestors(T key, Node<T> element, StringBuilder sb) {
+    if (element != null) {
+      int comparision = element.getKey().compareTo(key);
+
+      if (comparision == 0) {
+        //signal back that element is found and hence the recursion path back are the ancestors
+        return true;
+      } else if (printAncestors(key, element.getLeft(), sb) || printAncestors(key,
+          element.getRight(), sb)) {
+        sb.append(sb.length() > 0 ? ", " : "");
+        sb.append(element.getKey().toString());
+
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /* Question 2: Design an algorithm and write code to find the lowest common ancestor of two nodes
+                 in a binary tree. Avoid storing additional nodes in a data structure
+
+     Overview: The solution is implemented recursively.
+
+               Possible alternatives (omitted duals) considered in the recursive commonAncestor method:
+               1) None of the element are in the tree (else branch in commonAncestor)
+               2) One of the element is in the tree (bothFound is not set to true)
+               3) key1 is an ancestor of key2 => key2 (else if branch)
+               4) key1 is descendant on the left of some node and key2 is a descendant on thr right of the same node (if branch)
+
+      Time: O(N) where N is the number of elements in tree
+   */
+  T commonAncestor(T key1, T key2) {
+    bothFound = false;
+    T res = commonAncestor(key1, key2, root);
+
+    return bothFound ? res : null;
+  }
+
+  private T commonAncestor(T key1, T key2, Node<T> element) {
```

Comment:
I think we can assume that the keys in the tree are unique. However, should we consider cases like

- 17
   - 15
       - 8
   - 8
       - 15

Atm, this is not handled my approach?